proposals/0061-fair-congestion-control-with-intra-block-exponential-local-base-fee.md

```diff
@@ -0,0 +1,127 @@
+
+Related proposals:
+
+(TODO: add any relation of this to them)
+
+dynamic base fees
+https://github.com/solana-foundation/solana-improvement-documents/pull/4
+
+program rebatable account write fees:
+https://github.com/solana-foundation/solana-improvement-documents/pull/16
+
+asynchronous program execution:
+https://github.com/solana-foundation/solana-improvement-documents/pull/45
+
+increase prioritization fee:
+https://github.com/solana-foundation/solana-improvement-documents/pull/50
+
+bankless
+https://github.com/solana-foundation/solana-improvement-documents/pull/5
+
+## New Terminology
+
+Is there any new terminology introduced with this proposal?
+
+casual tx:
+fairness:
+block fullness in terms of number of actively-execution threads
+
+## Detailed Design
+
+(i jotted this down in 10min before going to bed! pardon for being so random
+writings...)
+
+to *determiniscally* define active thread count (`TC_a`), additionally record
+transaction termination events into poh stream.
+
+also, derive stake-weighted average transaction execution thread count
+(`TC_stake_weighted`).
+
+so, full is defined as the duration when  `TC_a == TC_stake_weighted` (this is
+updated at ~10ms intervals).
+
+when not full, maximize throughput of each of any single threaded transaction
+executions. note that, this mode exponentially cools down any hot addresses if
+any.
+
+when full, effectively pause any txes touching the hot state by exponentially
+increasing the local base fees. so casual txes can be executed.
+
+so, leaders are incentivised to manipulate in this naive form.
+
+so, split priority fee into two parts: (1) collected, (2) accrued for the next
+tx's base fee payment. The portion of (1) is calculated as if tx's cu *
+`TC_stake_weighted`. (i.e. as if validator stuffed spam txes to capture the
+exessive part (2) of prirotiy fee)
+
+in this way, there's no meaning to spam blocks by leaders. at the same time,
+it's still incentivied to pack txes according to p.f. desceding order, because
+leaders and clients alike are want to increse of their single threaded tps)
+
+also requested fee is basis for fee cals, block fullness calc, not the actual
+cu.
+- to prevent bad behavior, rebate 50% of (requested CU - actual CU)?
+  - so that leaders want txes success (want to increase actual CU) => usually
+    execute in the order
+  - specified by user
+  - so that users want txes fail fast (want to decrease actual CU)
+  - 25% is burntd and 25% is collected to leaders
+    - so, avoid too much requested cu.
+
+finally, when substantial blocks are full for extended duration, the global base-fee
+will naturally starts ceiling up. That's unavoidable no matter what.
+
+## Impact
+
+How will the implemented proposal impacts dapp developers, validators, and core contributors?
+
+## Security Considerations
+
+What security implications/considerations come with implementing this feature?
+Are there any implementation-specific guidance or pitfalls?
```

Comment:
this creates opportunity for a new type of MEV where an actor purposefully schedules transactions in a way that creates larger base fees. mango's validator could try to increase the fees for zeta to gain monetary benefit through attracting traders with lower fees while using another exchange.

Reply:
hey, thanks for jumping in.

the actual base fee calculation will be defined so that these arbitrary re-ordering of txes won't result in higher base fees, and larger total (base + priority) fees.

so, given 10 zeta honest transactions, if naive scheduling results in 1000 lamports are burnt, any reordering also results in 1000 lamports.

the base fee in this proposal is intended to throttle particularly hot address's tx throughput as the first objective.

also, tx specifies fixed base fee and fixed priority fee and it's 100% paid.

Reply:
you can still re-order across slots

Reply:
no worries. i haven't got time to write details, but i think this proposal can address this just like intra-block malicious reorderings.

as a teaser, a general idea is that base fee is reset across block boundaries unlike eip-1559. and raising base fees like that won't help because these hot addresses are forced to throttle executions according to new consensus logic. so, best strategy for leaders is spread out hot txes as much as possible across ledger ticks.